crypto/bio/bio_lib.c

```diff
@@ -786,41 +786,48 @@ void bio_cleanup(void)
 }
 
 /* Internal variant of the below BIO_wait() not calling BIOerr() */
-static int bio_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
+static int bio_wait(BIO *bio, time_t max_time, unsigned int polling_millisec)
 {
 #ifndef OPENSSL_NO_SOCK
     int fd;
 #endif
+    long sec_diff;
 
-    if (max_time == 0)
+    if (max_time == 0) /* no timeout */
         return 1;
 
 #ifndef OPENSSL_NO_SOCK
-    if (BIO_get_fd(bio, &fd) > 0)
+    if (BIO_get_fd(bio, &fd) > 0 && fd < FD_SETSIZE)
```

Comment:
Actually the strategy to fall back to polling can
work for read/write but wont work for connect.
Reason is that in the non-blocking case, you have
to wait until the socket is in writable state, but when
the select can't be used for that, you can't re-try
the connect, but have to use getsockopt(s, SOL_SOCKET, SO_ERROR, ...)
and that returns != 0 if an error happened,
but 0 when the connect succeeds, but also when the connect is
not yet completed.  You can assume the connect succeeds,
and continue with send/receive data which will fail, but we dont
have the data to send, and instead want to try the next
address if the previous one fails.

Reply:
It would make the bio_wait function more useful if it tries
to use the poll(2) API if it is available, and just fail if it is
not available.  I would recommend not to fall back the sleep
and also not to expose this API at all when NO_SOCK is in effect.

Reply:
> Actually the strategy to fall back to polling can
> work for read/write but wont work for connect.
> Reason is that in the non-blocking case, you have
> to wait until the socket is in writable state, but when
> the select can't be used for that, you can't re-try
> the connect, but have to use getsockopt(s, SOL_SOCKET, SO_ERROR, ...)
> and that returns != 0 if an error happened,
> but 0 when the connect succeeds, but also when the connect is
> not yet completed. You can assume the connect succeeds,
> and continue with send/receive data which will fail, but we dont
> have the data to send, and instead want to try the next
> address if the previous one fails.

> It would make the bio_wait function more useful if it tries
> to use the poll(2) API if it is available, and just fail if it is
> not available. I would recommend not to fall back the sleep
> and also not to expose this API at all when NO_SOCK is in effect.

I've tried several times to understand what exactly you meant here.
When resorting in the non-blocking case to polling then this is because the BIO is not socket-based and then one likely cannot use `getsockopt()` nor the poll API since they work on a socket.
For non-socket based connects it should still make sense to nap and poll again on a failed connect - AFAICS you don't have other chances than doing that (or simply failing potentially prematurely).

Please note also that `BIO_wait()` is also used for read/write, and at least there we agree that falling back to polling does make sense.